cumulus_library/apis/umls.py

```diff
@@ -0,0 +1,152 @@
+"""Class for communicating with the umls API"""
+import os
+import pathlib
+
+import requests
+
+from cumulus_library import base_utils, errors
+
+VALID_UMLS_DOWNLOADS = [
+    "rxnorm-full-monthly-release",
+    "rxnorm-weekly-updates",
+    "rxnorm-prescribable-content-monthly-release",
+    "rxnorm-prescribable-content-weekly-updates",
+    "rxnav-in-a-box",
+    "snomed-ct-us-edition",
+    "snomed-ct-us-edition-transitive-closure-resources",
+    "snomed-ct-international-edition",
+    "snomed-ct-core-problem-list-subset",
+    "snomed-ct-to-icd-10-cm-mapping-resources",
+    "snomed-ct-spanish-edition",
+    "umls-metathesaurus-full-subset",
+    "umls-metathesaurus-mrconso-file",
+    "umls-full-release",
+]
+
+
+class UmlsApi:
+    def __init__(self, api_key: str | None = None, validator_key: str | None = None):
+        """Creates a requests session for future calls, and validates the API key
+
+        :keyword api_key: A UMLS API key (will check for an env var named UMLS_API_KEY
+            if None)
+        :keyword validator_key: A UMLS API key for the calling application. Will be
+            set to the value of api_key if None (which is the current expected
+            behavior, since we don't want to be distributing UMLS keys)
+        """
+
+        if api_key is None:
+            api_key = os.environ.get("UMLS_API_KEY", None)
+            if api_key is None:
+                raise errors.ApiError("No UMLS API key provided")
+        self.api_key = api_key
+        if validator_key is None:
+            validator_key = api_key
+        self.validator_key = validator_key
+
+        auth_payload = {"validatorApiKey": self.validator_key, "apiKey": self.api_key}
+        self.session = requests.Session()
+        response = self.session.get(
+            "https://utslogin.nlm.nih.gov/validateUser", params=auth_payload
```

Comment:
basically my idea here was to deal with the bad key once, rather than having to catch it each time, so we don't have to do a bunch of different flavors of catching elsewhere.

I think this api set is maintained by multiple teams - some places ask for it as a basic auth token, where some ask for it as a request param of varying different names.

i could move this to a function call if you felt strongly about it, but i think this is the DRYest version.